editor/src/components/editor/actions/actions.tsx

```diff
@@ -5306,6 +5332,33 @@ export const UPDATE_FNS = {
 
     return updatedEditor
   },
+  UPDATE_TOP_LEVEL_ELEMENTS: (action: UpdateTopLevelElements, editor: EditorModel): EditorModel => {
+    const updatedEditor = modifyParseSuccessAtPath(
+      action.fullPath,
+      editor,
+      (parsed) => {
+        const newTopLevelElementsDeepEquals = arrayDeepEquality(TopLevelElementKeepDeepEquality)(
+          parsed.topLevelElements,
+          action.topLevelElements,
+        )
+
+        if (newTopLevelElementsDeepEquals.areEqual) {
+          return parsed
+        } else {
+          return {
+            ...parsed,
+            topLevelElements: newTopLevelElementsDeepEquals.value,
+          }
+        }
+      },
+      false,
+    )
+
+    return {
+      ...updatedEditor,
+      filesModifiedByAnotherUser: updatedEditor.filesModifiedByAnotherUser.concat(action.fullPath),
```

Comment:
something about it doesn't feel right. I think the action's name should be more specific, and contain Collab, or Remote, or some other word indicating that we are updating the top level elements here because an update came from the internet

Reply:
Yeah I think that's a good point. It doesn't feel right to you because from the action name it's not at all obvious that there is some logic in here that will always run but in fact should only run as a result of a collaborative editing change. I think at this point simply renaming the action is enough, but this is partly why we created `actions-from-vscode.ts`

Reply:
Fixed in ea4a8cf0e.